TOY-11 카메라 타이머 기능 개발

pby2017 opened the pull request:
코드 리뷰 부탁드립니다.

* [github / CamTimer, 타이머 코드 참고](https://github.com/dozingcat/CamTimer)
* 타이머 버튼 기능 정의
  * final list에 0, 2, 5, 10 int 값으로 초기화
  * 클릭시 setText로 0, 2, 5, 10 초 변경
  * timer delay 멤버 변수에 시간 값 저장
* 사진 촬영 버튼 클릭시 타이머 기능 정의
  * timer delay 시간값 조건문 확인 후 Handler로 count down 시간 띄워주기
  * count down 끝나면 사진 촬영 수행

pby2017 committed:
TOY-11 카메라 타이머 기능 개발

bonchangkoo committed:
Merge branch 'develop' into TOY-11-camera-timer

bonchangkoo reviewed (changes requested):
제 테스트 환경에서는 타이머 기능이 아예 정상 동작 하지 않네요. 확인 부탁드려요.

pby2017 committed:
TOY-11 btn_capture 클릭 이벤트 merge 내용 수정

pby2017 commented:
@bonchangkoo 님
> 제 테스트 환경에서는 타이머 기능이 아예 정상 동작 하지 않네요. 확인 부탁드려요.

동작할 수 있도록 수정했습니다.
코드 리뷰 부탁드립니다.

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/camera/CameraActivity.java:
```diff
@@ -102,11 +127,15 @@ protected void onStart() {
     protected void onPause() {
         cameraKitView.onPause();
+        if (timerHandler != null) {
+            finishDelayCapture();
```
직접 Debug 해보니까 타이머 카운트 도중에 백키 누르면 두번 해당 함수가 두번 호출되네요.

pby2017 replied:
@bonchangkoo 님,
중복 확인은 코드 리뷰 하면서 확인했습니다.

이렇게 코드를 작성했던 의도는
1. onBackPressed() 의 경우
  * 타이머 작동 중간에 뒤로가기 버튼을 누르면 액티비티는 유지하고 타이머만 취소되도록 하려고 했습니다.
  * 하지만 현재 의도한바와 다르게, 타이머 작동  중간에 뒤로가기를 누르면 액티비티가 종료됩니다.
2. onPause() 의 경우
  * 촬영 후 미리보기 화면으로 넘어갈 때 타이머로 인해 VISIBILITY 처리되었던 카운트다운 텍스트를 GONE 처리하려고 했습니다.

bonchangkoo replied:
타이머가 동작하지 않을 때 timerHandler null 이 아닐 수 있을 거 같네요.

pby2017 replied:
@bonchangkoo 님,
timeHandler null 상태 체크 관련해서 확인하고 수정했습니다.
코드 리뷰 부탁드립니다.

> 타이머가 동작하지 않을 때 timerHandler null 이 아닐 수 있을 거 같네요.

타이머를 설정하지 않아도 timeHandler는 null로 설정하는 코드가 없었기 때문에 항상 null 이 아니었습니다.
그 부분에 대해서 timeHandler가 null 인지 체크하는 코드를 서브 스레드 큐에 타이머 관련 작업이 남아있는지 체크하는 코드로 수정했습니다.

그리고 타이머가 취소되면 removeCallbacksAndMessages() 메서드를 호출해서 서브 스레드의 작업을 제거하도록 처리했습니다.

pby2017 committed:
TOY-11 timeHandler null 상태 체크 관련 수정

타이머를 설정하지 않아도 timeHandler는 null로 설정하는 코드가 없었기 때문에 항상 null 이 아니었습니다.
그 부분에 대해서 timeHandler가 null 인지 체크하는 코드를 서브 스레드 큐에 타이머 관련 작업이 남아있는지 체크하는 코드로 수정했습니다.

그리고 타이머가 취소되면 removeCallbacksAndMessages() 메서드를 호출해서 서브 스레드의 작업을 제거하도록 처리했습니다.

bonchangkoo reviewed (approved):
코드 확인했습니다.

hyejin830 commented:
타이머 실행 이후 프리뷰액티비티에서 뒤로 갈 때 타이머 상태 그대로입니다. 확인부탁드립니다~

pby2017 committed:
TOY-11 timer 작동시 컨트롤 버튼 Visibility 설정 관련 수정

기존 onResume() 에 컨트롤 버튼 Visibility를 설정해주는 코드가 있었는데
단순, 백그라운드 복귀 프로세스 같은 경우 불필요한 코드 호출이 있을 것 같고,
onResume()은 잘 생각하고 정의하셔야 한다는 조언에 따라서

최근 수정한 코드에서는 Visibility 코드를 제거하고,
대신 타이머를 정상적으로 또는 강제적으로 끝낼 때 호출하는 finishDelayCapture() 메서드 안에 Visibility 코드를 정의했었습니다.

다만, Handler 안의 큐를 모두 삭제하는 코드가 먼저 호출이 되어
타이머 작동 후 촬영이 끝나고 호출되어야 하는 finishDelayCapture() 가 무시되었던 점에서 문제가 발생했습니다.

이번에는 onResume()이 아닌 타이머 작동 후 촬영 메서드를 호출하는 코드 다음에 Visibility를 설정하는 코드를 추가했습니다.

pby2017 commented:
@hyejin830 님,
timer 작동시 컨트롤 버튼 Visibility 설정 관련해서 확인하고 수정했습니다.
코드 리뷰 부탁드립니다.

> 타이머 실행 이후 프리뷰액티비티에서 뒤로 갈 때 타이머 상태 그대로입니다. 확인부탁드립니다~

기존 onResume() 에 컨트롤 버튼 Visibility를 설정해주는 코드가 있었는데
단순, 백그라운드 복귀 프로세스 같은 경우 불필요한 코드 호출이 있을 것 같고,
onResume()은 잘 생각하고 정의하셔야 한다는 조언에 따라서

최근 수정한 코드에서는 onResume()에서 Visibility 설정 코드를 제거하고,
대신 타이머를 정상적으로 또는 강제적으로 끝낼 때 호출하는 finishDelayCapture() 메서드 안에 Visibility 코드를 정의했었습니다.

다만, Handler 안의 큐를 모두 삭제하는 코드가 먼저 호출이 되어
타이머 작동 후 촬영이 끝나고 호출되어야 하는 finishDelayCapture() 가 무시되었던 점에서 
컨트롤 버튼의 Visibility 설정 또한 무시되는 문제가 발생했습니다.

이번에는 onResume()이 아닌 타이머 작동 후 촬영 메서드를 호출하는 코드 다음에 Visibility를 설정하는 코드를 추가했습니다.

hyejin830 commented:
확인했습니다~

pby2017 committed:
Merge branch 'develop' into TOY-11-camera-timer

# Conflicts:
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/camera/CameraActivity.java
#	app/src/main/res/layout/activity_camera.xml
